### PR TITLE
Scrubbing

### PR DESCRIPTION
**Pull request by KrishPatel13**
Presidio is not capable for Scrubbing Credit-Card numbers and Social Security Numbers (ssn), which is strange (it should do it). See Test file and the attached test session text file for reference.
Besides that, it is working for the below fields:

- Email
- Phone Number
- Date of Birth
- Address
- Driving Licence
- Passport
- National ID
- Routing Number
- Bank Account Number
-

### CODE REVIEW

**Comment by abrichr** [comment]
Excellent work @KrishPatel13 !

It looks like the scrubbing is working, and the test needs to be modified:

```assert '************...n **********.' == "****'s email...n **/**/****."```

**Comment by KrishPatel13** [comment]
My next step will be Implementing scrubbing for images

File: tests/puterbot/test_scrub.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import pytest
+from puterbot.scrub import scrub
+
```
Comment (abrichr) [inline tests/puterbot/test_scrub.py:3]: Please add two newlines after imports

File: tests/puterbot/test_scrub.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+def test_empty_string():
+    text = ""
+    expected_output = ""
+    assert scrub(text) == expected_output
+
```
Comment (abrichr) [inline tests/puterbot/test_scrub.py:8]: Please add two newlines between function definitions

File: requirements.txt [hunk 3]
```diff
@@ -22,3 +22,7 @@ sqlalchemy==1.4.43
 torch==2.0.0
 tqdm==4.64.0
 transformers==4.28.1
+presidio_analyzer==2.2.32
+presidio_anonymizer==2.2.32
+presidio_image_redactor==0.0.46
+
```
Comment (abrichr) [inline requirements.txt:28]: Please remove this empty newline

File: puterbot/scrub.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+# Required Packages: 
+# pip install presidio_analyzer
+# pip install presidio_anonymizer
+# pip install presidio_image_redactor
+# Note: they have been added in requirement.txt
+# python -m spacy download en_core_web_lg (after installing the required packages)
+
+def scrub(text: str) -> str:
+    """
+        This function takes in a string of text as input and returns a scrubbed version of the text with any potential personally
```
Comment (abrichr) [inline puterbot/scrub.py:16]: Thank you for the docstring!

Please follow the format described here: https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: puterbot/scrub.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        text=text,
+        analyzer_results=analyzer_results,     # type: ignore
+        operators=operators
+    )
+
+    return anonymized_results.text
+
+
+if __name__ == '__main__':
+    import sys
```
Comment (abrichr) [inline puterbot/scrub.py:46]: Please move imports to the top of the file

File: puterbot/scrub.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    inp_text = "My credit card number is 1234-5678-9012-3456 and "
+    scrubbed_text = scrub(inp_text)
+    print(scrubbed_text)
```
Comment (abrichr) [inline puterbot/scrub.py:49]: Please use `loguru.logger`

File: puterbot/scrub.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    # Note: This script uses the sys.stdin.read() method to read input from the console,
```
Comment (abrichr) [inline puterbot/scrub.py:51]: Please move usage instructions to the top of the file

File: tests/puterbot/test_scrub.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    assert scrub("My bank routing number is 123456789.") == "My bank routing number is *********."
+
+def test_scrub_bank_account():
+    # Test scrubbing of bank account number
+    assert scrub("My bank account number is 123456789012.") == "My bank account number is ************."
+
+def test_scrub_all_together():
+    # Text with all PII/PHI types
+    text_with_pii_phi = "John Smith's email is johnsmith@example.com and his phone number is 555-123-4567. His credit card number is 1234-5678-9012-3456 and his social security number is 123-45-6789. He was born on 01/01/1980."
+    assert scrub(text_with_pii_phi) == "****'s email is ***@***.*** and his phone number is ***-***-****. His credit card number is ****-****-****-**** and his social security number is ***-**-****. He was born on **/**/****."
```
Comment (abrichr) [inline tests/puterbot/test_scrub.py:61]: Please make sure lines are under 80 characters in length; see https://github.com/psf/black

**Comment by KrishPatel13** [comment]
On more research, I figured out that, Presidio indeed scrubs Credit Card Numbers. 

The reason it was not scrubbing the test case "my credit card number is 1234-1234-1234" is that, it is NOT a VALID Credit-card number.

Explanation: Looking at the codebase for presidio repo of Microsoft. Their code also check for IIN (Issuer Identification Numbers),

Examples of issuer identification numbers include the following: ([Reference](https://www.investopedia.com/terms/i/issuer-identification-number-iin.asp#:~:text=The%20first%20six%20to%20eight,card%20network%E2%80%94not%20the%20cardholder.))
- American Express: 34, 37
- Discover Card: 6011, 622126 to 622925, 624000 to 626999, 628200 to 628899, 64, 65
- Mastercard: 2221 to 2720, 51 to 55
- Visa: 4

This is reason why the number in the test case 1234-1234-1234-1234 was not scrubbing.

On further testing, it successfully scurbs  the Visa Credit card numbers, that are, starting from 4.

File: puterbot/scrub.py (hunk 6 above)
Comment (KrishPatel13) [reply puterbot/scrub.py]: I was just testing here. I am done with the testing now, so I have removed this lines now.

**Comment by KrishPatel13** [comment]
Re-review requested for current Text based scrubbing module: scrub.py

Next-Steps:

Moving on with Image Scrubbing.

File: puterbot/scrub.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    """
+    analyzer = AnalyzerEngine()
+    analyzer_results = analyzer.analyze(
+        text=text, entities=analyzer.get_supported_entities(), language="en")
+    anonymizer = AnonymizerEngine()
+
+    operators = {}
+    for entity in analyzer_results:
+        operators[entity.entity_type] = OperatorConfig(
+            "mask", {"masking_char": "*", "chars_to_mask": MAX_MASK_LEN, "from_end": True})
```
Comment (abrichr) [inline puterbot/scrub.py:35]: Please make sure lines are less than 80 chars.

See https://github.com/psf/black

File: tests/puterbot/test_scrub.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+    """ This test is to ensure that an empty string is returned 
+    if an empty string is passed to the scrub function.
+    
+    Args:
+        None
+    Returns:
+        None
+    Raises: 
+        None
+    """
```
Comment (abrichr) [inline tests/puterbot/test_scrub.py:15]: No need to document like this for tests 👍

File: tests/puterbot/test_scrub.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+    text = ""
+    expected_output = ""
+    assert scrub(text) == expected_output
+
+
+def test_no_scrub_string() -> None:
+    """
+    This test is to ensure that the same string is returned
```
Comment (abrichr) [inline tests/puterbot/test_scrub.py:23]: Please word it as "Test to ensure..."

File: tests/puterbot/test_scrub.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+    Args:
+        None
+    Returns:
+        None
+    Raises: 
+        None
+    """
+    # Test scrubbing of email address
+    assert scrub(
+        "My email is john.doe@example.com.") == "My email is ********************."
```
Comment (abrichr) [inline tests/puterbot/test_scrub.py:50]: Please format symmetrically, e.g.:
```
    assert scrub(
        "My email is john.doe@example.com."
    ) == "My email is ********************."
```

**Review by abrichr — COMMENTED** [review]
Just a few small comments, looks good!

File: tests/puterbot/test_scrub.py (hunk 10 above)
Comment (KrishPatel13) [reply tests/puterbot/test_scrub.py:15]: noted!

**Comment by KrishPatel13** [comment]
Re-review requested

**Comment by KrishPatel13** [comment]
@abrichr
https://github.com/MLDSAI/OpenAdapt/issues/47#issuecomment-1544041785

**Comment by KrishPatel13** [comment]
I am getting this error. On line 23 of [test_bash_image.sh](https://github.com/MLDSAI/OpenAdapt/pull/128/commits/1a4992b3ab2493b2e7e1d3cc43c22b106b6d9918#diff-7ee217e22b775c563958c0ee53da0d42b8a8c97ec266bd9ecf5f8c31e6d6abe7)

My goal is to delete/remove the scrubbed_image.png file after the test.

**Comment by KrishPatel13** [comment]
@abrichr  Rereview requested 

- [ ] Step 1: Scrub Text                
- [ ] Step 2: Scrub Image Data
- [ ] Step 3: [write_input_event](https://github.com/MLDSAI/puterbot/blob/main/puterbot/record.py#L148) is trickier because a single input event may not contain PII (e.g. the letter "b"), but a sequence of them will (e.g. the sequence "bob smith")
InputEvents are merged in [get_consecutive_keyboard_events](https://github.com/MLDSAI/puterbot/blob/main/puterbot/events.py#L389). This logic is nontrivial and currently incomplete, and may run several times before everything is merged.

The merging process is completed after the process loop is completed in [get_events](https://github.com/MLDSAI/puterbot/blob/main/puterbot/events.py#L69). This is probably the better place to implement scrubbing of InputEvents.
- [ ] Step 4: Edit record.py and call the appropriate function for scrubbing at different places. As listed down in [here](https://github.com/MLDSAI/OpenAdapt/issues/47#issuecomment-1537213925).

**Comment by KrishPatel13** [comment]
@abrichr 

Should I call the scrubbing module before Only Visualization or before the replay as well ?
Alternatively, should I just save the scrubbed recording/events into the db. No matter. Which approach seems better ?

**Comment by KrishPatel13** [comment]
@abrichr 

To scrub the text from the Window Event, I would need the text from it. I saw the Window Event class. But no where it has a field of text. how can I access it ?

Will this be available once this issue https://github.com/MLDSAI/OpenAdapt/issues/75 will be resolved! ? If yes, how would name the the new text field for the Window Event ? If you could tell me then, I can continue assuming I have the access to that field so that I can call scrub function aptly ? How does this sound ?

**Comment by KrishPatel13** [comment]
@abrichr
Since my record was not still working, I asked Aaron to run my code on his MacOS
The changes made before that, before visualization, I just called the scrub_image function and then displayed each screen shot.
Here is what he got! Any thoughts? 

https://we.tl/t-K1e6bLjoLc[https://we.tl/t-K1e6bLjoLc](url)

**Comment by KrishPatel13** [comment]
@abrichr 

On further inspection, I do thing we should ignore URL as well because, wherever it detect a dot '.' It scrubs that. What do you think ?

```
SCRUB_IGNORE_ENTITIES = [
    # 'US_PASSPORT',
    # 'US_DRIVER_LICENSE',
    # 'CRYPTO',
    # 'UK_NHS',
    # 'PERSON',
    # 'CREDIT_CARD',
    # 'US_BANK_NUMBER',
    # 'PHONE_NUMBER',
    # 'US_ITIN',
    # 'AU_ABN',
    "DATE_TIME",
    # 'NRP',
    # 'SG_NRIC_FIN',
    # 'AU_ACN',
    # 'IP_ADDRESS',
    # 'EMAIL_ADDRESS',
    # 'URL',
    # 'IBAN_CODE',
    # 'AU_TFN',
    # 'LOCATION',
    # 'AU_MEDICARE',
    # 'US_SSN',
    # 'MEDICAL_LICENSE'
]
```

**Comment by KrishPatel13** [comment]
Eg. See Before and After Scrubbing Images below

After:
image_720

Before: 
image_720

**Comment by KrishPatel13** [comment]
@abrichr 

Ready for Re-View.

**Comment by KrishPatel13** [comment]
Some Notes from the Sprint Retrospective Meeting: 
> Customized URL detection: Though we are NOT scrubbing all the URLs right now. But we want to scrub some specific URLs. 
> Scrubb the action events as well. The Key shown on the Screen shots eg. `<M+y+ + c+r+d+i+t+ +c+a+r+d+ n+u+m+b+e+r+ +i+s+ +4+5+0+1+-+2+3+4+5+-+9+8+7+0+.>`
So we would like it to scrub to something like:
`<M+y+ + c+r+d+i+t+ +c+a+r+d+ n+u+m+b+e+r+ +i+s+ +*+*+*+*+*+*+*+*+*+*+*+*+*+*+*>`

File: tests/openadapt/test_scrub.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
+"""Module to test scrub.py"""
+from io import BytesIO
+import os
+from PIL import Image
```
Comment (abrichr) [inline tests/openadapt/test_scrub.py:4]: Newline before third party modules

File: tests/openadapt/test_scrub.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+    # Save the scrubbed image data to a file
+    scrubbed_image_path = "scrubbed_image.png"
+    scrubbed_image.save(scrubbed_image_path)
+
+    # Load the scrubbed image from file for manual verification
+    scrubbed_image = Image.open(scrubbed_image_path)
+
+    # Perform OCR on the scrubbed image
+    ocr_text = pytesseract.image_to_string(scrubbed_image)
```
Comment (abrichr) [inline tests/openadapt/test_scrub.py:32]: As I understand it, this test is a smoke test to make sure that the module runs, as there is no automated analysis of the results.

For manual verification like you are doing here, there should be something like this at the bottom of `scrub.py`:

```
if __name__ == "__main__":
    fire.Fire(utils.get_functions(__name__))
```
utils.get_functions:
```
def get_functions(name):
    """
    Get a dictionary of function names to functions for all non-private functions

    Usage:

        if __name__ == "__main__":
            fire.Fire(utils.get_functions(__name__))

    Args:
        name: string containing name of module in sys.modules

    Returns:
        A dictionary of function names to functions.
    """
    functions = {}
    for name, obj in inspect.getmembers(sys.modules[name]):
        if inspect.isfunction(obj) and not name.startswith("_"):
            functions[name] = obj
    return functions
```

For this test, you can either:
1. remove this line and the one before it, and call it a "smoke" test
2. add some functionality to ensure that the output is as expected, e.g. with PIL by counting the number of pixels having the color of the mask and asserting it is e.g. 1/10th the total number of pixels in the image (or whatever is appropriate)

File: tests/openadapt/test_scrub.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+
+    test_image.close()
+    scrubbed_image.close()
+    os.remove(scrubbed_image_path)
+
+    assert "krish@openadapt.ai" not in ocr_text
```
Comment (abrichr) [inline tests/openadapt/test_scrub.py:38]: Excellent

File: tests/openadapt/test_scrub.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+    assert "Manage your Google Account" in ocr_text
+    assert "Sign out" in ocr_text
+
+
+def test_empty_string() -> None:
+    """
+    Test is to ensure that an empty string is returned
+    if an empty string is passed to the scrub function.
+    """
+    text = ""
```
Comment (abrichr) [inline tests/openadapt/test_scrub.py:48]: Please make sure you have newlines after docstrings elsewhere, I will not continue to point this out in this PR

File: tests/openadapt/test_scrub.py [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+    expected_output = ""
+    assert scrub(text) == expected_output
+
+
+def test_no_scrub_string() -> None:
+    """
+    Test is to ensure that the same string is returned
```
Comment (abrichr) [inline tests/openadapt/test_scrub.py:55]: From https://stackoverflow.com/questions/59902102/why-is-imperative-mood-important-for-docstrings:
> First line should be in imperative mood.

@Mustaballer can we please use pydocstyle?

**Review by abrichr — CHANGES_REQUESTED** [review]
Excellent work @KrishPatel13 ! Just a few small changes and this will be good to merge 👍

File: openadapt/scrub.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    Returns:
+        str: Scrubbed text
+
+    Raises:
+        None
+    """
+    analyzer_results = analyzer.analyze(
+        text=text, entities=SCRUBBING_ENTITIES, language="en"
+    )
+    anonymizer = AnonymizerEngine()
```
[the remark this answers is not in the document]
Comment (KrishPatel13) [reply openadapt/scrub.py:40]: Because analyzer is needed to both of the function. So in order to reduce code redundancy I made analyzer global.

File: openadapt/scrub.py [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+        bytes: Scrubbed PNG data
+
+    Raises:
+        None
+    """
+    # Initialize the engine
+    engine = ImageRedactorEngine()
+
+    # Redact the image with red color
+    redacted_image = engine.redact(image, fill=(255,), entities=SCRUBBING_ENTITIES)
```
[the remark this answers is not in the document]
Comment (KrishPatel13) [reply openadapt/scrub.py:76]: Nice Idea! Adding it :)

File: openadapt/scrub.py [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+    analyzer_results = analyzer.analyze(
+        text=text, entities=SCRUBBING_ENTITIES, language="en"
+    )
+    anonymizer = AnonymizerEngine()
+
+    operators = {}
+    for entity in analyzer_results:
+        operators[entity.entity_type] = OperatorConfig(
+            "mask",
+            {"masking_char": "*", "chars_to_mask": MAX_MASK_LEN, "from_end": True},
```
[the remark this answers is not in the document]
Comment (KrishPatel13) [reply openadapt/scrub.py:46]: I ran it it automatically make it in one line. 
I manually made it short, it automatically puts all of them together 😅

image

[the remark this answers is not in the document]
Comment (KrishPatel13) [reply openadapt/scrub.py:46]: Got it working!

image

File: tests/openadapt/test_scrub.py (hunk 16 above)
Comment (KrishPatel13) [reply tests/openadapt/test_scrub.py:48]: Noted!

File: tests/openadapt/test_scrub.py (hunk 17 above)
Comment (KrishPatel13) [reply tests/openadapt/test_scrub.py:55]: Done for test_scrub.py

File: tests/openadapt/test_scrub.py (hunk 14 above)
Comment (KrishPatel13) [reply tests/openadapt/test_scrub.py]: Picked Option 2. Looked more robust.!

**Review by KrishPatel13 — COMMENTED** [review]
Reviewed and Changed!

**Comment by KrishPatel13** [comment]
@abrichr Re-review is requested!

Thank you for your patience!

**Review by KrishPatel13 — COMMENTED** [review]
Reviewed

**Comment by abrichr** [comment]
Thank you @KrishPatel13 ! Can you please merge the latest from main?

File: README.md [hunk 21]
```diff
@@ -214,4 +215,4 @@ Please submit any issues to https://github.com/MLDSAI/openadapt/issues with the
 following information:
 
 - Problem description (please include any relevant console output and/or screenshots)
-- Steps to reproduce (please help others to help you!)
+- Steps to reproduce (please help others to help you!)
```
Comment (abrichr) [inline README.md:218]: Please remove unrelated changes 🙏

**Review by abrichr — CHANGES_REQUESTED** [review]

**Comment by KrishPatel13** [comment]
https://github.com/microsoft/presidio/issues/1079

**Comment by KrishPatel13** [comment]
> [microsoft/presidio#1079](https://github.com/microsoft/presidio/issues/1079)
> 
> I was not able to record so I finally had to investigate deep and found that after adding simplest code of Presidio of initializing the engines it gave the errors. So I had to raise the issue on their GitHub page.

Resolved:

Issue Description: 

If I pressed Ctrl + C after 3 "starting". It gave errors. I even did some typing and then pressed Ctrl + C, got the error.

BUT,

when I waited for atleast 10-15sec idle after the 3 starting and then press ctrl + C, it recorded successfully.

**Comment by KrishPatel13** [comment]
@abrichr We can close this WITHOUT merging once the below is merged! :)

https://github.com/MLDSAI/OpenAdapt/pull/211
Ready for review! :)

**Comment by abrichr** [comment]
Looks great @KrishPatel13 ! Can you please look into modifying configure_logging to scrub data before it is logged? If it's a lot of work we can leave it as a TODO for now.

**Comment by abrichr** [comment]
Closing in favor of https://github.com/MLDSAI/OpenAdapt/pull/211
